### PR TITLE
`CompileLoss`: fix for partially defined loss with different `y_pred` and `y_true` structures

### PR DESCRIPTION
**Pull request by nicolaspi**
Fix for cases where the loss is partially defined over `y_pred`.

Example:
```python

inputs = keras.Input((2,))
x = keras.layers.Dense(3)(inputs)
model = keras.Model(inputs, [x, x, x])
model.compile(loss=["mse", None, None])

model.fit(x=np.random.random((320, 2)), y=np.random.random((320, 3)))
```

### CODE REVIEW

File: keras/src/trainers/compile_utils_test.py [hunk 1]
```diff
@@ -520,3 +520,15 @@ def test_structure_mismatch(self):
...
+        y_pred = [np.random.random((320, 3))] * 3
+        y_true = np.random.random((320, 3))
+
+        compile_loss = CompileLoss(
+            loss=["mse", None, None], output_names=["a", "b", "c"]
+        )
+        # build call
+        compile_loss(y_true, y_pred)
+        # built call
+        compile_loss(y_true, y_pred)
```
Comment (fchollet) [inline keras/src/trainers/compile_utils_test.py:534]: Please check that `model.evaluate()` is correct (so that we know we're computing MSE for the first input and not something else, like 1 MSE per input...)

e.g.

```python
inputs = layers.Input((2,))
x = layers.Dense(3, kernel_initializer="ones", use_bias=False)(inputs)
model = models.Model(inputs, [x, x, x])
model.compile(loss=["mse", None, None])
model.fit(x=np.ones((320, 2)), y=np.zeros((320, 3)))
out = model.evaluate(x=np.ones((320, 2)), y=np.zeros((320, 3)))
self.assertAllClose(out, 3.862943)
```

**Review by fchollet — COMMENTED** [review]
Thanks for the fix!

File: keras/src/trainers/compile_utils.py [hunk 2]
```diff
@@ -683,8 +693,22 @@ def call(self, y_true, y_pred, sample_weight=None):
...
+                            y_pred, tree.flatten(y_true)
+                        )
+                    except:
+                        # Check case where loss is partially defined over y_pred
+                        flat_y_true = tree.flatten(y_true)
+                        flat_loss = tree.flatten(self._user_loss)
+                        flat_loss_non_nones = [
+                            loss for loss in flat_loss if loss is not None
+                        ]
+                        assert len(flat_y_true) == len(flat_loss_non_nones)
```
[the remark this answers is not in the document]
Comment (nicolaspi) [reply keras/src/trainers/compile_utils.py:708]: This `assert` is encapsulated in a `try...except` block and follows the pattern of the preceding `assert_same_structure/paths` checks.
Shall I raise an exception there anyway?

[the remark this answers is not in the document]
Comment (fchollet) [reply keras/src/trainers/compile_utils.py:708]: Got it, sounds fine then!

File: keras/src/trainers/compile_utils_test.py (hunk 1 above)
Comment (nicolaspi) [reply keras/src/trainers/compile_utils_test.py]: I added a test that checks this.

**Review by fchollet — APPROVED** [review]
LGTM, thank you
